src/index.js

```diff
@@ -0,0 +1,135 @@
+'use strict'
+
+const { NOISE } = require('@chainsafe/libp2p-noise')
+const libp2p = require('libp2p')
+const Multiplex = require('libp2p-mplex')
+const Websockets = require('libp2p-websockets')
+const { blocksTable, getPeerId, primaryKeys, port } = require('./config')
+const { logger, serializeError } = require('./logging')
+const {
+  Message,
+  BITSWAP_V_100,
+  BITSWAP_V_110,
+  BITSWAP_V_120,
+  Block,
+  BlockPresence,
+  Entry,
+  WantList
+} = require('./message')
+const { receiveData, sendData } = require('./networking')
+const { cidToKey, fetchS3Object, readDynamoItem } = require('./storage')
+
+const emptyWantList = new WantList([], true)
+
+async function fetchBlock(cid) {
+  const info = await getBlockInfo(cid)
+
+  if (!info) {
+    return null
+  }
+
+  const { offset, length, car } = info.cars[0]
+  const separator = car.indexOf('/')
+  const bucket = car.slice(0, separator)
+  const key = car.slice(separator + 1)
+
+  return fetchS3Object(bucket, key, offset, length)
+}
+
+async function getBlockInfo(cid) {
+  const item = await readDynamoItem(blocksTable, primaryKeys.blocks, cidToKey(cid))
+  return item
+}
+
+async function processWantlist(stream, protocol, wantlist) {
+  const blocks = []
+  const presences = []
+
+  // For each entry in the list
+  for (const entry of wantlist.entries) {
+    // We don't care about canceling wantlist since we don't maintain state
+    if (entry.cancel) {
+      continue
+    }
+
+    if (entry.wantType === Entry.WantType.Block) {
+      // Fetch the block and eventually append to the list of blocks
+      const raw = await fetchBlock(entry.cid)
+
+      if (raw) {
+        blocks.push(new Block(entry.cid, raw))
+      } else if (entry.sendDontHave && protocol === BITSWAP_V_120) {
+        presences.push(new BlockPresence(entry.cid, BlockPresence.Type.DontHave))
+      }
+    } else if (entry.wantType === Entry.WantType.Have && protocol === BITSWAP_V_120) {
+      // Check if we have the block
+      const existing = await getBlockInfo(entry.cid)
+
+      if (existing) {
+        presences.push(new BlockPresence(entry.cid, BlockPresence.Type.Have))
+      } else if (entry.sendDontHave) {
+        presences.push(new BlockPresence(entry.cid, BlockPresence.Type.DontHave))
+      }
+    }
+  }
+
+  // pendingBytes is always 0 as we don't maintain a state
+  return new Message(emptyWantList, blocks, presences, 0)
```

Comment:
how much memory could this object allocate at worst?

Reply:
Is up to us. presences is usually small as each entry is a `{cid: byte[~40], type: 0 | 1}` object.
blocks are made of entry each up to 2 MB.

The maximum message size can be of 4 MB. - I just got this reply from PL and I still have to insert checks in the code.

Reply:
I would add this as docs.

Reply:
Sure. I will.

Reply:
Done. Will be in the next push.